src/array_api_extra/_apply.py

```diff
@@ -0,0 +1,285 @@
+"""Public API Functions."""
+
+# https://github.com/scikit-learn/scikit-learn/pull/27910#issuecomment-2568023972
+from __future__ import annotations
+
+from collections.abc import Callable, Sequence
+from functools import wraps
+from types import ModuleType
+from typing import TYPE_CHECKING, Any, cast, overload
+
+from ._lib._compat import (
+    array_namespace,
+    is_dask_namespace,
+    is_jax_namespace,
+)
+from ._lib._typing import Array, DType
+
+if TYPE_CHECKING:
+    # https://github.com/scikit-learn/scikit-learn/pull/27910#issuecomment-2568023972
+    from typing import ParamSpec, TypeAlias
+
+    import numpy as np
+
+    NumPyObject: TypeAlias = np.ndarray[Any, Any] | np.generic  # type: ignore[no-any-explicit]
+    P = ParamSpec("P")
+else:
+    # Sphinx hacks
+    NumPyObject = Any
+
+    class P:  # pylint: disable=missing-class-docstring
+        args: tuple
+        kwargs: dict
+
+
+@overload
+def apply_numpy_func(  # type: ignore[valid-type]
+    func: Callable[P, NumPyObject],
+    *args: Array,
+    shape: tuple[int, ...] | None = None,
+    dtype: DType | None = None,
+    xp: ModuleType | None = None,
+    **kwargs: P.kwargs,  # pyright: ignore[reportGeneralTypeIssues]
+) -> Array: ...  # numpydoc ignore=GL08
+
+
+@overload
+def apply_numpy_func(  # type: ignore[valid-type]
+    func: Callable[P, Sequence[NumPyObject]],
+    *args: Array,
+    shape: Sequence[tuple[int, ...]],
+    dtype: Sequence[DType] | None = None,
+    xp: ModuleType | None = None,
+    **kwargs: P.kwargs,  # pyright: ignore[reportGeneralTypeIssues]
+) -> tuple[Array, ...]: ...  # numpydoc ignore=GL08
+
+
+def apply_numpy_func(  # type: ignore[valid-type]  # numpydoc ignore=GL07,SA04
+    func: Callable[P, NumPyObject | Sequence[NumPyObject]],
+    *args: Array,
+    shape: tuple[int, ...] | Sequence[tuple[int, ...]] | None = None,
+    dtype: DType | Sequence[DType] | None = None,
+    xp: ModuleType | None = None,
+    **kwargs: P.kwargs,  # pyright: ignore[reportGeneralTypeIssues]
+) -> Array | tuple[Array, ...]:
+    """
+    Apply a function that operates on NumPy arrays to Array API compliant arrays.
+
+    Parameters
+    ----------
+    func : callable
+        The function to apply. It must accept one or more NumPy arrays or generics as
+        positional arguments and return either a single NumPy array or generic, or a
+        tuple or list thereof.
+
+        It must be a pure function, i.e. without side effects such as disk output,
+        as depending on the backend it may be executed more than once.
```

Comment:
This last line is a bit cryptic here, perhaps better to stop after "without side effects".